File: Java/Top_K_Frequent_Elements.java
public class Solution {
    class NumCount implements Comparable<NumCount> {
        int number;
        int count;
        public NumCount(int number, int count) {
            this.number = number;
            this.count = count;
        }
        public int compareTo(NumCount nc) {
            return count - nc.count;
        } 
    }
    
    public List<Integer> topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> m = new HashMap<Integer, Integer>();
        Integer tmp = null;
        for (int x : nums) {
            tmp = m.get(x);
            if (tmp == null) m.put(x, 1);
            else m.put(x, tmp+1);
        }
        PriorityQueue<NumCount> pq = new PriorityQueue<NumCount>();
        for (Map.Entry<Integer, Integer> e : m.entrySet()) {
            pq.add(new NumCount(e.getKey(), e.getValue()));
            if (pq.size() > k) pq.poll();
        }
        List<Integer> ret = new LinkedList<Integer>();
        for (NumCount nc : pq) {
            ret.add(nc.number);
        }
        return ret;
    }
}

File: Go/Top_K_Frequent_Elements.go
// Priority Queue Implementation
type NumCount struct {
    number int
    count int
    index int
}

type PriorityQueue []*NumCount

func (pq PriorityQueue) Len() int { return len(pq)}

func (pq PriorityQueue) Less(i,j int) bool{
    return pq[i].count > pq[j].count
}
func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *PriorityQueue) Push(x interface{}) {
	n := len(*pq)
	item := x.(*NumCount)
	*pq = append(*pq, item)
	item.index = n
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	item.index = -1 // for safety
	*pq = old[0 : n-1]
	return item
}
func topKFrequent(nums []int, k int) []int {
    m := make(map[int]int)
    for _,v := range nums {
        m[v] ++
    }
    pq := make(PriorityQueue, len(m))
    i := 0
    for k,v := range m {
        pq[i] = &NumCount{number: k, count: v, index: i}
        i ++
    }
    heap.Init(&pq)
    ret := make([]int, k)
    ret_ptr := 0
    for pq.Len() > 0 && k > 0 {
        ret[ret_ptr] = heap.Pop(&pq).(*NumCount).number
        ret_ptr++
        k --
    }
    return ret
}
